language/move-prover/bytecode-to-boogie/src/spec_translator.rs

```diff
@@ -0,0 +1,317 @@
+                } else {
+                    // Interpret as reference which we dref.
+                    // TODO: we may want to obtimize some cases here to avoid constructing
+                    //   an intermediate Reference.
+                    format!(
+                        "Dereference(m, {})",
+                        self.translate_location_as_reference(other)
+                    )
+                }
+            }
+        }
+    }
+
+    /// Translate a location as Reference.
+    fn translate_location_as_reference(&mut self, loc: &StorageLocation) -> String {
+        match loc {
+            StorageLocation::Formal(name) => {
+                format!("GetLocalReference(m_size, {})", self.get_param_index(name))
+            }
+            StorageLocation::GlobalResource { type_, address } => format!(
+                "GetResourceReference(gs, {}, {})",
+                self.translate_location_as_address(address),
+                self.translate_resource_name(type_)
+            ),
+            StorageLocation::AccessPath { base, fields } => {
+                let mut res = self.translate_location_as_reference(base);
+                for f in fields {
+                    res.push_str(&format!(
+                        "SelectField({}, {})",
+                        res,
+                        self.translate_field(f)
+                    ));
+                }
+                res
+            }
+            _ => self.error(
+                &format!("cannot translate as reference: {:?}", loc),
+                "<ref>".to_string(),
+            ),
+        }
+    }
+
+    /// Translate a location as Address.
+    fn translate_location_as_address(&mut self, addr: &StorageLocation) -> String {
+        match addr {
+            StorageLocation::TxnSenderAddress => "TxnSenderAddress()".into(),
+            StorageLocation::Address(addr) => self.translate_account_address(addr),
+            _ => self.error(
+                &format!("cannot translate as address: {:?}", addr),
+                "<addr>".to_string(),
+            ),
+        }
+    }
+
+    /// Checks whether this is a translatable address.
+    fn is_address_location(&self, addr: &StorageLocation) -> bool {
+        match addr {
+            StorageLocation::TxnSenderAddress => true,
+            StorageLocation::Address(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Translate an account address literal.
+    fn translate_account_address(&self, addr: &AccountAddress) -> String {
+        format!("{}", BigInt::from_str_radix(&addr.to_string(), 16).unwrap())
+    }
+
+    /// Translates a resource name.
+    fn translate_resource_name(&mut self, name: &StructName) -> String {
+        // StructName is (at least now) parsed as a simple, non-qualified name. As there can be
+        // no qualification, simple names should be sufficient to identify the full struct info
```

Comment:
This seems quite reasonable for now, but there is potential ambiguity for:
- Structs with the same name declared in different modules
- The above + modules with the same name published under different addresses

Probably worth adding a TODO and/or a check that makes the verifier fail if the `StructHandle`s of the module under analysis can cause ambiguity.

Reply:
Added TODO